File: back-end/src/main/java/com/example/backend/service/TripService.java
package com.example.backend.service;

import com.example.backend.dto.trip.TripRequest;
import com.example.backend.entities.Trip;
import com.example.backend.service.base.BaseService;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public interface TripService extends BaseService<Trip> {
    List<Trip> findByDriverId(Long driverId);
    List<Trip> findByVehicleId(Long vehicleId);
    List<Trip> findByRouteId(Long routeId);
    List<Trip> findUpcomingTrips(LocalDateTime start, LocalDateTime end);
    List<Trip> findActiveTrips();
    List<Trip> findUpcomingTripsByDriver(Long driverId);
    long countActiveTripsForSchool(Long schoolId);
    void assignDriver(Long tripId, Long driverId);
    void assignVehicle(Long tripId, Long vehicleId);
    void updateTripStatus(Long tripId, String status);
    void updateTripSchedule(Long tripId, LocalDateTime departureTime, LocalDateTime estimatedArrivalTime);
    List<Trip> findTripsWithStats(Long schoolId, LocalDateTime start, LocalDateTime end);
    void startTrip(Long tripId);
    void completeTrip(Long tripId);
    void cancelTrip(Long tripId, String reason);
    Trip findById(Long id);
    List<Trip> findByStudentId(Long studentId);
    List<Trip> findBySchoolId(Long schoolId);
    List<Trip> findByDateRange(LocalDateTime start, LocalDateTime end);
    List<Trip> findByStatus(String status);
    Trip save(Trip trip);
    void delete(Long id);
    Trip findByIdWithDetails(Long id);
    Trip createAndStartTrip(Long routeId);
    Trip endCurrentTrip();
    Trip getCurrentTrip();
    List<Trip> getTripHistory(int page, int size);
    Trip findCurrentTripByStudentId(Long studentId);
    List<Trip> findTripsByStudentId(Long studentId, int page, int size);
    Trip create(@Valid TripRequest tripRequest);
    Trip assignStudents(Long tripId, List<Long> studentIds);
} 

File: front-end/src/services/TripService.ts
import { Trip } from '../core/entities/trip.entity';

export interface TripService {
  getAllTrips(): Promise<Trip[]>;
  getTripById(id: number): Promise<Trip>;
  getTripWithDetails(id: number): Promise<Trip>;
  createTrip(tripData: any): Promise<Trip>;
  deleteTrip(id: number): Promise<void>;
  findByDriverId(driverId: number): Promise<Trip[]>;
  findByVehicleId(vehicleId: number): Promise<Trip[]>;
  findByRouteId(routeId: number): Promise<Trip[]>;
  findUpcomingTrips(start: string, end: string): Promise<Trip[]>;
  findActiveTrips(): Promise<Trip[]>;
  findUpcomingTripsByDriver(driverId: number): Promise<Trip[]>;
  countActiveTripsForSchool(schoolId: number): Promise<number>;
  assignDriver(tripId: number, driverId: number): Promise<void>;
  assignVehicle(tripId: number, vehicleId: number): Promise<void>;
  assignRoute(tripId: number, routeId: number): Promise<void>;
  updateTripStatus(tripId: number, status: string): Promise<void>;
  updateTripSchedule(tripId: number, departureTime: string, estimatedArrivalTime: string): Promise<void>;
  findTripsWithStats(schoolId: number, start: string, end: string): Promise<Trip[]>;
  startTrip(tripId: number): Promise<void>;
  completeTrip(tripId: number): Promise<void>;
  cancelTrip(tripId: number, reason: string): Promise<void>;
  findByStudentId(studentId: number): Promise<Trip[]>;
  findBySchoolId(schoolId: number): Promise<Trip[]>;
  findByDateRange(start: string, end: string): Promise<Trip[]>;
  findByStatus(status: string): Promise<Trip[]>;
  createAndStartTrip(routeId: number): Promise<Trip>;
  endCurrentTrip(): Promise<Trip>;
  getCurrentTrip(): Promise<Trip>;
  getTripHistory(page: number, size: number): Promise<Trip[]>;
  findCurrentTripByStudentId(studentId: number): Promise<Trip>;
  findTripsByStudentId(studentId: number, page: number, size: number): Promise<Trip[]>;
  assignStudents(tripId: number, studentIds: number[]): Promise<Trip>;
  getTripRequestSchema(): Promise<any>;
} 